usr/src/man/man1/chown.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited All Rights Reserved
.\" Portions Copyright (c) 1982-2007 AT&T Knowledge Ventures
.\" Copyright (c) 2008, Sun Microsystems, Inc. All Rights Reserved
.\"
.TH CHOWN 1 "Feb 21, 2019"
.SH NAME
chown \- change file ownership
.SH SYNOPSIS
.LP
.nf
\fBchown\fR [\fB-fhR\fR] \fIowner\fR[:\fIgroup\fR] \fIfile\fR...
.fi

.LP
.nf
\fBchown\fR \fB-s\fR [\fB-fhR\fR] \fIownersid\fR[:\fIgroupsid\fR] \fIfile\fR...
.fi

.LP
.nf
\fBchown\fR \fB-R\fR [\fB-f\fR] [\fB-H\fR | \fB-L\fR | \fB-P\fR] \fIowner\fR[:\fIgroup\fR] \fIfile\fR...
.fi

.LP
.nf
\fBchown\fR \fB-s\fR \fB-R\fR [\fB-f\fR] [\fB-H\fR | \fB-L\fR | \fB-P\fR] \fIownersid\fR[:\fIgroupsid\fR] \fIfile\fR...
.fi

.SH DESCRIPTION
.LP
The \fBchown\fR utility sets the user \fBID\fR of the file named by each
\fBfile\fR to the user \fBID\fR specified by \fIowner\fR, and, optionally, sets
the group \fBID\fR to that specified by \fIgroup\fR.
.sp
.LP
If \fBchown\fR is invoked by other than the super-user, the set-user-\fBID\fR
bit is cleared.
.sp
.LP
Only the owner of a file (or the super-user) can change the owner of that file.
.sp
.LP
The operating system has a configuration option
\fB{_POSIX_CHOWN_RESTRICTED}\fR, to restrict ownership changes. When this
option is in effect the owner of the file is prevented from changing the owner
\fBID\fR of the file. Only the super-user can arbitrarily change owner
\fBID\fRs whether or not this option is in effect. To set this configuration
option, include the following line in \fB/etc/system\fR:
.sp
.in +2
.nf
\fBset rstchown = 1\fR
.fi
.in -2
.sp

.sp
.LP
To disable this option, include the following line in \fB/etc/system\fR:
.sp
.in +2
.nf
\fBset rstchown = 0\fR
.fi
.in -2
.sp

.sp
.LP
\fB{_POSIX_CHOWN_RESTRICTED}\fR is enabled by default. See \fBsystem\fR(5) and
\fBfpathconf\fR(2).
.SH OPTIONS
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-f\fR\fR
.ad
.RS 6n
Force. Does not report errors.
.RE

.sp
.ne 2
.na
\fB\fB-h\fR\fR
.ad
.RS 6n
If the file is a symbolic link, this option changes the owner of the symbolic
link. Without this option, the owner of the file referenced by the symbolic
link is changed.
.RE

.sp
.ne 2
.na
\fB\fB-H\fR\fR
.ad
.RS 6n
If the file specified on the command line is a symbolic link referencing a file
of type directory, this option changes the owner of the directory referenced by
the symbolic link and all the files in the file hierarchy below it. If a
symbolic link is encountered when traversing a file hierarchy, the owner of the
target file is changed, but no recursion takes place.
.RE

.sp
.ne 2
.na
\fB\fB-L\fR\fR
.ad
.RS 6n
If the file is a symbolic link, this option changes the owner of the file
referenced by the symbolic link. If the file specified on the command line, or
encountered during the traversal of the file hierarchy, is a symbolic link
referencing a file of type directory, then this option changes the owner of the
directory referenced by the symbolic link and all files in the file hierarchy
below it.
.RE

.sp
.ne 2
.na
\fB\fB-P\fR\fR
.ad
.RS 6n
If the file specified on the command line or encountered during the traversal
of a file hierarchy is a symbolic link, this option changes the owner of the
symbolic link. This option does not follow the symbolic link to any other part
of the file hierarchy.
.RE

.sp
.ne 2
.na
\fB\fB-R\fR\fR
.ad
.RS 6n
Recursive. \fBchown\fR descends through the directory, and any subdirectories,
setting the specified ownership \fBID\fR as it proceeds. When a symbolic link
is encountered, the owner of the symbolic link is changed, unless the
\fB-H\fR or \fB-L\fR option is specified. Unless the \fB-H\fR, \fB-L\fR, or \fB-P\fR
option is specified, the \fB-P\fR option is used as the default mode.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR\fR
.ad
.RS 6n
The owner and/or group arguments are Windows SID strings. This option requires
a file system that supports storing SIDs, such as ZFS.
.RE

.sp
.LP
Specifying more than one of the mutually-exclusive options \fB-H\fR, \fB-L\fR,
or \fB-P\fR is not considered an error. The last option specified determines
the behavior of \fBchown\fR.

.SH OPERANDS
.LP
The following operands are supported:
.sp
.ne 2
.na
\fB\fIowner\fR[\fB:\fR\fIgroup\fR]\fR
.ad
.RS 17n
A user \fBID\fR and optional group \fBID\fR to be assigned to \fBfile\fR. The
\fIowner\fR portion of this operand must be a user name from the user database
or a numeric user \fBID\fR. Either specifies a user \fBID\fR to be given to
each file named by \fIfile\fR. If a numeric \fIowner\fR exists in the user
database as a user name, the user \fBID\fR number associated with that user
name is used as the user \fBID\fR. Similarly, if the \fIgroup\fR portion of
this operand is present, it must be a group name from the group database or a
numeric group \fBID\fR. Either specifies a group \fBID\fR to be given to each
file. If a numeric group operand exists in the group database as a group name,
the group \fBID\fR number associated with that group name is used as the group
\fBID\fR.
.RE

.sp
.ne 2
.na
\fB\fIfile\fR\fR
.ad
.RS 17n
A path name of a file whose user \fBID\fR is to be modified.
.RE

.SH USAGE
.LP
See \fBlargefile\fR(7) for the description of the behavior of \fBchown\fR when
encountering files greater than or equal to 2 Gbyte ( 2^31 bytes).
.SH EXAMPLES
.LP
\fBExample 1 \fRChanging Ownership of All Files in the Hierarchy
.sp
.LP
The following command changes ownership of all files in the hierarchy,
including symbolic links, but not the targets of the links:

.sp
.in +2
.nf
example% \fBchown \(miR \(mih \fIowner\fR[:group] \fIfile\fR...\fR
.fi
.in -2
.sp

.SH ENVIRONMENT VARIABLES
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBchown\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 6n
The utility executed successfully and all requested changes were made.
.RE

.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
An error occurred.
.RE

.SH FILES
.ne 2
.na
\fB\fB/etc/passwd\fR\fR
.ad
.RS 15n
System password file
.RE

.SH ATTRIBUTES
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
CSI	Enabled. See \fBNOTES\fR.
_
Interface Stability	Committed
_
Standard	See \fBstandards\fR(7).
.TE

.SH SEE ALSO
.LP
.BR chgrp (1),
.BR chmod (1),
.BR ksh93 (1),
.BR chown (2),
.BR fpathconf (2),
.BR passwd (5),
.BR system (5),
.BR attributes (7),
.BR environ (7),
.BR largefile (7),
.BR standards (7)
.SH NOTES
.LP
\fBchown\fR is \fBCSI\fR-enabled except for the \fIowner\fR and \fIgroup\fR
names.
.sp
.LP
In the past the behavior of \fB/usr/xpg4/bin/chown\fR and
\fB/usr/bin/chown\fR utilities was different. Now they behave the same way.


usr/src/cmd/chown/chown.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2008 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1983, 1984, 1985, 1986, 1987, 1988, 1989 AT&T	*/
/*	All Rights Reserved	*/

/*
 * Portions of this source code were derived from Berkeley 4.3 BSD
 * under license from the Regents of the University of California.
 */

/*
 * chown [-fhR] uid[:gid] file ...
 * chown -R [-f] [-H|-L|-P] uid[:gid] file ...
 * chown -s [-fhR] ownersid[:groupsid] file ...
 * chown -s -R [-f] [-H|-L|-P] ownersid[:groupsid] file ...
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <sys/types.h>
#include <dirent.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/avl.h>
#include <pwd.h>
#include <grp.h>
#include <unistd.h>
#include <locale.h>
#include <errno.h>
#include <libcmdutils.h>
#include <aclutils.h>

static struct		passwd	*pwd;
static struct		group	*grp;
static struct		stat	stbuf;
static uid_t		uid = (uid_t)-1;
static gid_t		gid = (gid_t)-1;
static int		status = 0;	/* total number of errors received */
static int		hflag = 0,
			rflag = 0,
			fflag = 0,
			Hflag = 0,
			Lflag = 0,
			Pflag = 0,
			sflag = 0;
static avl_tree_t	*tree;

static int		Perror(char *);
static int		isnumber(char *);
static void		chownr(char *, uid_t, gid_t);
static void		usage();

/*
 * Check to see if we are to follow symlinks specified on the command line.
 * This assumes we've already checked to make sure neither -h or -P was
 * specified, so we are just looking to see if -R -H, or -R -L was specified.
 */
#define	FOLLOW_CL_LINKS	(rflag && (Hflag || Lflag))

/*
 * Follow symlinks when traversing directories.  Only follow symlinks
 * to other parts of the file hierarchy if -L was specified.
 */
#define	FOLLOW_D_LINKS	(Lflag)

#define	CHOWN(f, u, g)	if (chown(f, u, g) < 0) { \
				status += Perror(f); \
			}
#define	LCHOWN(f, u, g)	if (lchown(f, u, g) < 0) { \
				status += Perror(f); \
			}


int
main(int argc, char *argv[])
{
	int c;
	int ch;
	char *grpp;			/* pointer to group name arg */
	extern int optind;
	int errflg = 0;

	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)		/* Should be defined by cc -D */
#define	TEXT_DOMAIN	"SYS_TEST"	/* Use this only if it weren't */
#endif
	(void) textdomain(TEXT_DOMAIN);

	while ((ch = getopt(argc, argv, "hRfHLPs")) != EOF) {
		switch (ch) {
		case 'h':
			hflag++;
			break;

		case 'R':
			rflag++;
			break;

		case 'f':
			fflag++;
			break;

		case 'H':
			/*
			 * If more than one of -H, -L, and -P
			 * are specified, only the last option
			 * specified determines the behavior of
			 * chown.
			 */
			Lflag = Pflag = 0;
			Hflag++;
			break;

		case 'L':
			Hflag = Pflag = 0;
			Lflag++;
			break;

		case 'P':
			Hflag = Lflag = 0;
			Pflag++;
			break;

		case 's':
			sflag++;
			break;

		default:
			errflg++;
			break;
		}
	}
	/*
	 * Set Pflag by default for recursive operations
	 * if no other options were specified.
	 */
	if (rflag && !(Lflag || Hflag || Pflag || hflag)) {
		Pflag = 1;
	}
	/*
	 * Check for sufficient arguments
	 * or a usage error.
	 */

	argc -= optind;
	argv = &argv[optind];

	if (errflg || (argc < 2) ||
	    ((Hflag || Lflag || Pflag) && !rflag) ||
	    ((Hflag || Lflag || Pflag) && hflag)) {
		usage();
	}

	/*
	 * POSIX.2
	 * Check for owner[:group]
	 */
	if ((grpp = strchr(argv[0], ':')) != NULL) {
		*grpp++ = 0;

		if (sflag) {
			if (sid_to_id(grpp, B_FALSE, &gid)) {
				(void) fprintf(stderr, gettext(
				    "chown: invalid owning group sid %s\n"),
				    grpp);
				exit(2);
			}
		} else if ((grp = getgrnam(grpp)) != NULL) {
			gid = grp->gr_gid;
		} else {
			if (isnumber(grpp)) {
				errno = 0;
				gid = (gid_t)strtoul(grpp, NULL, 10);
				if (errno != 0) {
					if (errno == ERANGE) {
						(void) fprintf(stderr, gettext(
						"chown: group id too large\n"));
							exit(2);
					} else {
						(void) fprintf(stderr, gettext(
						"chown: invalid group id\n"));
						exit(2);
					}
				}
			} else {
				(void) fprintf(stderr, gettext(
				    "chown: unknown group id %s\n"), grpp);
				exit(2);
			}
		}
	}

	if (sflag) {
		if (sid_to_id(argv[0], B_TRUE, &uid)) {
			(void) fprintf(stderr, gettext(
			    "chown: invalid owner sid %s\n"), argv[0]);
			exit(2);
		}
	} else if ((pwd = getpwnam(argv[0])) != NULL) {
		uid = pwd->pw_uid;
	} else {
		if (isnumber(argv[0])) {
			errno = 0;
			uid = (uid_t)strtoul(argv[0], NULL, 10);
			if (errno != 0) {
				if (errno == ERANGE) {
					(void) fprintf(stderr, gettext(
					"chown: user id too large\n"));
					exit(2);
				} else {
					(void) fprintf(stderr, gettext(
					"chown: invalid user id\n"));
					exit(2);
				}
			}
		} else {
			(void) fprintf(stderr, gettext(
			"chown: unknown user id %s\n"), argv[0]);
			exit(2);
		}
	}

	for (c = 1; c < argc; c++) {
		tree = NULL;
		if (lstat(argv[c], &stbuf) < 0) {
			status += Perror(argv[c]);
			continue;
		}
		if (rflag && ((stbuf.st_mode & S_IFMT) == S_IFLNK)) {
			if (hflag || Pflag) {
				/*
				 * Change the ownership of the symlink
				 * specified on the command line.
				 * Don't follow the symbolic link to
				 * any other part of the file hierarchy.
				 */
				LCHOWN(argv[c], uid, gid);
			} else {
				struct stat stbuf2;
				if (stat(argv[c], &stbuf2) < 0) {
					status += Perror(argv[c]);
					continue;
				}
				/*
				 * We know that we are to change the
				 * ownership of the file referenced by the
				 * symlink specified on the command line.
				 * Now check to see if we are to follow
				 * the symlink to any other part of the
				 * file hierarchy.
				 */
				if (FOLLOW_CL_LINKS) {
					if ((stbuf2.st_mode & S_IFMT)
					    == S_IFDIR) {
						/*
						 * We are following symlinks so
						 * traverse into the directory.
						 * Add this node to the search
						 * tree so we don't get into an
						 * endless loop.
						 */
						if (add_tnode(&tree,
						    stbuf2.st_dev,
						    stbuf2.st_ino) == 1) {
							chownr(argv[c],
							    uid, gid);
						} else {
							/*
							 * Error occurred.
							 * rc can't be 0
							 * as this is the first
							 * node to be added to
							 * the search tree.
							 */
							status += Perror(
							    argv[c]);
						}
					} else {
						/*
						 * Change the user ID of the
						 * file referenced by the
						 * symlink.
						 */
						CHOWN(argv[c], uid, gid);
					}
				} else {
					/*
					 * Change the user ID of the file
					 * referenced by the symbolic link.
					 */
					CHOWN(argv[c], uid, gid);
				}
			}
		} else if (rflag && ((stbuf.st_mode & S_IFMT) == S_IFDIR)) {
			/*
			 * Add this node to the search tree so we don't
			 * get into a endless loop.
			 */
			if (add_tnode(&tree, stbuf.st_dev,
			    stbuf.st_ino) == 1) {
				chownr(argv[c], uid, gid);
			} else {
				/*
				 * An error occurred while trying
				 * to add the node to the tree.
				 * Continue on with next file
				 * specified.  Note: rc shouldn't
				 * be 0 as this was the first node
				 * being added to the search tree.
				 */
				status += Perror(argv[c]);
			}
		} else if (hflag || Pflag) {
			LCHOWN(argv[c], uid, gid);
		} else {
			CHOWN(argv[c], uid, gid);
		}
	}
	return (status);
}

/*
 * chownr() - recursive chown()
 *
 * Recursively chowns the input directory then its contents.  rflag must
 * have been set if chownr() is called.  The input directory should not
 * be a sym link (this is handled in the calling routine).  In
 * addition, the calling routine should have already added the input
 * directory to the search tree so we do not get into endless loops.
 * Note: chownr() doesn't need a return value as errors are reported
 * through the global "status" variable.
 */
static void
chownr(char *dir, uid_t uid, gid_t gid)
{
	DIR *dirp;
	struct dirent *dp;
	struct stat st, st2;
	char savedir[1024];

	if (getcwd(savedir, 1024) == (char *)0) {
		(void) Perror("getcwd");
		exit(255);
	}

	/*
	 * Attempt to chown the directory, however don't return if we
	 * can't as we still may be able to chown the contents of the
	 * directory.  Note: the calling routine resets the SUID bits
	 * on this directory so we don't have to perform an extra 'stat'.
	 */
	CHOWN(dir, uid, gid);

	if (chdir(dir) < 0) {
		status += Perror(dir);
		return;
	}
	if ((dirp = opendir(".")) == NULL) {
		status += Perror(dir);
		return;
	}
	for (dp = readdir(dirp); dp != NULL; dp = readdir(dirp)) {
		if (strcmp(dp->d_name, ".") == 0 ||	/* skip . and .. */
		    strcmp(dp->d_name, "..") == 0) {
			continue;
		}
		if (lstat(dp->d_name, &st) < 0) {
			status += Perror(dp->d_name);
			continue;
		}
		if ((st.st_mode & S_IFMT) == S_IFLNK) {
			if (hflag || Pflag) {
				/*
				 * Change the ownership of the symbolic link
				 * encountered while traversing the
				 * directory.  Don't follow the symbolic
				 * link to any other part of the file
				 * hierarchy.
				 */
				LCHOWN(dp->d_name, uid, gid);
			} else {
				if (stat(dp->d_name, &st2) < 0) {
					status += Perror(dp->d_name);
					continue;
				}
				/*
				 * We know that we are to change the
				 * ownership of the file referenced by the
				 * symlink encountered while traversing
				 * the directory.  Now check to see if we
				 * are to follow the symlink to any other
				 * part of the file hierarchy.
				 */
				if (FOLLOW_D_LINKS) {
					if ((st2.st_mode & S_IFMT) == S_IFDIR) {
						/*
						 * We are following symlinks so
						 * traverse into the directory.
						 * Add this node to the search
						 * tree so we don't get into an
						 * endless loop.
						 */
						int rc;
						if ((rc = add_tnode(&tree,
						    st2.st_dev,
						    st2.st_ino)) == 1) {
							chownr(dp->d_name,
							    uid, gid);
						} else if (rc == 0) {
							/* already visited */
							continue;
						} else {
							/*
							 * An error occurred
							 * while trying to add
							 * the node to the tree.
							 */
							status += Perror(
							    dp->d_name);
							continue;
						}
					} else {
						/*
						 * Change the user id of the
						 * file referenced by the
						 * symbolic link.
						 */
						CHOWN(dp->d_name, uid, gid);
					}
				} else {
					/*
					 * Change the user id of the file
					 * referenced by the symbolic link.
					 */
					CHOWN(dp->d_name, uid, gid);
				}
			}
		} else if ((st.st_mode & S_IFMT) == S_IFDIR) {
			/*
			 * Add this node to the search tree so we don't
			 * get into a endless loop.
			 */
			int rc;
			if ((rc = add_tnode(&tree, st.st_dev,
			    st.st_ino)) == 1) {
				chownr(dp->d_name, uid, gid);
			} else if (rc == 0) {
				/* already visited */
				continue;
			} else {
				/*
				 * An error occurred while trying
				 * to add the node to the search tree.
				 */
				status += Perror(dp->d_name);
				continue;
			}
		} else {
			CHOWN(dp->d_name, uid, gid);
		}
	}

	(void) closedir(dirp);
	if (chdir(savedir) < 0) {
		(void) fprintf(stderr, gettext(
		    "chown: can't change back to %s\n"), savedir);
		exit(255);
	}
}

static int
isnumber(char *s)
{
	int c;

	while ((c = *s++) != '\0')
		if (!isdigit(c))
			return (0);
	return (1);
}

static int
Perror(char *s)
{
	if (!fflag) {
		(void) fprintf(stderr, "chown: ");
		perror(s);
	}
	return (!fflag);
}

static void
usage()
{
	(void) fprintf(stderr, gettext(
	    "usage:\n"
	    "\tchown [-fhR] owner[:group] file...\n"
	    "\tchown -R [-f] [-H|-L|-P] owner[:group] file...\n"
	    "\tchown -s [-fhR] ownersid[:groupsid] file...\n"
	    "\tchown -s -R [-f] [-H|-L|-P] ownersid[:groupsid] file...\n"));
	exit(2);
}
